src/document.rs

```diff
@@ -1,27 +1,368 @@
-    _rev: String,
+    pub _rev: String,
+}
+
+/// Standard Response when having CRUD operation on document
+/// Todo could add more response variant like shown on official doc
+/// 200: is returned if the document was found
+/// 304: is returned if the “If-None-Match” header is given and the document has
+/// the same version 404: is returned if the document or collection was not
+/// found 412: is returned if an “If-Match” header is given and the found
+/// document has a different version. The response will also contain the found
+/// document’s current revision in the Etag header.
+pub enum DocumentResponse<T> {
+    /// Silent is when there is empty object returned by the server
+    Silent,
+    /// Contain data after CRUD
+    Response {
+        header: DocumentHeader,
+        old: Option<T>,
+        new: Option<T>,
+        _old_rev: Option<String>,
+    },
+    Err(ArangoError),
+}
+
+/// Gives extra method on the DocumentResponse to quickly check what the server
+/// returns
+impl<T> DocumentResponse<T> {
+    /// Should be true when the server send back an empty object {}
+    pub fn is_silent(&self) -> bool {
+        match self {
+            DocumentResponse::Silent => true,
+            _ => false,
+        }
+    }
+    /// Should be true if there is a response from the server
+    pub fn has_response(&self) -> bool {
+        match self {
+            DocumentResponse::Response { .. } => true,
+            _ => false,
+        }
+    }
+
+    /// Return the document header contained inside the response
+    pub fn header(&self) -> Option<&DocumentHeader> {
+        if let DocumentResponse::Response { header, .. } = self {
+            Some(header)
+        } else {
+            None
+        }
+    }
+    /// Return the old document before changes
+    pub fn old_doc(&self) -> Option<&T> {
+        Option::from(if let DocumentResponse::Response { old, .. } = self {
+            old
+        } else {
+            &None
+        })
+    }
+    /// Return the new document
+    pub fn new_doc(&self) -> Option<&T> {
+        Option::from(if let DocumentResponse::Response { new, .. } = self {
+            new
+        } else {
+            &None
+        })
+    }
+    /// return the old revision of the document
+    pub fn old_rev(&self) -> Option<&String> {
+        Option::from(if let DocumentResponse::Response { _old_rev, .. } = self {
+            _old_rev
+        } else {
+            &None
+        })
+    }
+}
 
+impl<'de, T> Deserialize<'de> for DocumentResponse<T>
+where
+    T: Deserialize<'de>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let mut obj = serde_json::Value::deserialize(deserializer)?;
+
+        if obj.get_mut("error").is_some() {
+            return ArangoError::deserialize(obj)
+                .map(DocumentResponse::Err)
+                .map_err(de::Error::custom);
+        }
```

Comment:
Don't need to handle `error`. This is already handle by `deserialize_response`. Only those without `error` can be passed on to the deseialization of  `DocumentResponse`

I think we can just delete this code block and let go of `DocumentResponse::Err`.

```rust
        if obj.get_mut("error").is_some() {
            return ArangoError::deserialize(obj)
                .map(DocumentResponse::Err)
                .map_err(de::Error::custom);
        }
```